src/libstore/downstream-placeholder.hh

```diff
@@ -0,0 +1,71 @@
+#pragma once
+///@file
+
+#include "hash.hh"
+#include "path.hh"
+
+namespace nix {
+
+/**
+ * Downstream Placeholders are opaque and almost certainly unique values
+ * used to allow derivations to refer to store objects which are yet to
+ * be built and for we do not yet have store paths for.
+ *
+ * They correspond to `DerivedPaths` that are not `DerivedPath::Opaque`,
+ * except for the cases involving input addressing or fixed outputs
+ * where we do know a store path for the derivation output in advance.
+ *
+ * Unlike `DerivationPath`, however, `DownstreamPlaceholder` is
+ * purposefully opaque and obfuscated. This is so they are hard to
+ * create by accident, and so substituting them (once we know what the
+ * path to store object is) is unlikely to capture other stuff it
+ * shouldn't.
+ *
+ * We use them with `Derivation`: the `render()` method is called to
+ * render an opaque string which can be used in the derivation, and the
+ * resolving logic can substitute those strings for store paths when
+ * resolving `Derivation.inputDrvs` to `BasicDerivation.inputSrcs`.
+ */
+class DownstreamPlaceholder
+{
+    /**
+     * `DownstreamPlaceholder` is just a newtype of `Hash`.
+     * This its only field.
+     */
+    Hash hash;
+
+public:
+    /**
+     * This creates an opaque and almost certainly unique string
+     * deterministically from the placeholder.
+     */
+    std::string render() const;
+
+    /**
+     * For content-addressed derivations
+     *
+     * The derivation is known (we have a store path for it), but the
+     * output doesn't yet have a known store path.
+     */
+    DownstreamPlaceholder(const StorePath & drvPath, std::string_view outputName);
+
+    /**
+     * For dynamic derivations
+     *
+     * The derivation is not yet known, and we just have (another)
+     * placeholder for it.
+     */
+    DownstreamPlaceholder(const DownstreamPlaceholder & drvPlaceholder, std::string_view outputName);
```

Comment:
A named constructor would be preferable. Overloading hides the distinction that happens here.